### PR TITLE
EVG-7819 activate/deactivate blocked dependencies

### PR DESCRIPTION
**Pull request by ybrill**
If a task depends on a task which has been deactivated it's essentially blocked. It should not come up for consideration every time the scheduler runs. It hasn't been included in how "blocked" has been defined until now, so I chose to exclude these tasks by deactivating them. If one of a task's dependencies is deactivated it will get deactivated. If the dependency is subsequently activated we will reactivate it (only if it was deactivated for this reason).

Along the way I consolidated the places where tasks are activated/deactivated to simplify task lifecycle (a little). This should also help logging discrepancies (it wasn't always logged to the event log when a task was activated/deactivated) and the ActivatedBy field will always be set.

If the DAG dispatcher finds a task we've missed it will deactivate dependencies. If it finds a blocked task (i.e. a task with a dependency that has finished running with a status that doesn't fulfill the dependency) it will mark it blocked (until now both cases were deactivated, causing [EVG-7819](https://jira.mongodb.org/browse/EVG-7819)).

The recursive calls shouldn't be too bad. The call stack will grow only as deep as the height of the dependency tree and we only do one(ish) db query per call for all the tasks. Deactivating a whole build/version will batch all its tasks together so we'll usually only make one recursive query per build/version (as opposed to one per task).

### CODE REVIEW

File: model/lifecycle.go [hunk 1]
```diff
@@ -82,80 +82,65 @@ func SetVersionActivation(versionId string, active bool, caller string) error {
...
-				return errors.Wrapf(err, "problem finding tasks with dependencies for builds %s", buildIds)
-			}
-			catcher := grip.NewBasicCatcher()
-			for _, t := range tasks {
-				for _, d := range t.DependsOn {
-					catcher.Add(SetActiveState(d.TaskId, caller, active))
-				}
-			}
-			grip.Error(errors.Wrapf(catcher.Resolve(), "problem settings dependencies for builds %s", buildIds))
+		dependOn, err := task.GetRecursiveDependenciesUp(tasks, nil)
```
Comment (ablack12) [inline model/lifecycle.go:110]: check error

File: model/task/task.go [hunk 2]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
 		bson.M{
-			IdKey: t.Id,
+			IdKey: bson.M{"$in": taskIDs},
 		},
 		bson.M{
 			"$set": bson.M{
 				ActivatedKey:     true,
 				ActivatedByKey:   caller,
-				ActivatedTimeKey: t.ActivatedTime,
+				ActivatedTimeKey: time.Now(),
```
Comment (ablack12) [inline model/task/task.go:905]: In the ActivateTask case, the local t.ActivatedTime and the DB ActivatedTime won't match, fyi.

File: model/task/task.go [hunk 3]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
 			},
 		})
+	if err != nil {
+		return errors.Wrap(err, "can't activate task")
```
Comment (ablack12) [inline model/task/task.go:909]: nit, tasks

File: model/task/task.go [hunk 4]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+	}
+
+	for _, t := range tasks {
+		event.LogTaskActivated(t.Id, t.Execution, caller)
```
Comment (ablack12) [inline model/task/task.go:913]: We should choose a convention about whether or not we log regardless of error, since DeactivateDependencies does this differently.

File: model/task/task.go [hunk 5]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+			continue
+		}
+
+		for _, dep := range t.DependsOn {
+			if !taskMap[dep.TaskId] && !depTaskMap[dep.TaskId] {
+				tasksToGet = append(tasksToGet, dep.TaskId)
+			}
+		}
+	}
+	missingTasks, err := FindAll(db.Query(bson.M{IdKey: bson.M{"$in": tasksToGet}}).WithFields(ActivatedKey))
```
Comment (ablack12) [inline model/task/task.go:949]: we can skip this if tasksToGet is empty

File: model/task/task.go [hunk 6]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+		if t.Activated || !t.DeactivatedForDependency {
+			continue
+		}
+
+		depsSatisfied := true
+		for _, dep := range t.DependsOn {
+			// not being activated now
+			if _, ok := tasksToActivate[dep.TaskId]; !ok && !taskMap[dep.TaskId] {
+				// and not already activated
+				if depTask := missingTaskMap[dep.TaskId]; !depTask.Activated {
```
Comment (ablack12) [inline model/task/task.go:975]: Should we also check ok here?

File: model/task/task.go [hunk 7]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+		bson.M{"$set": bson.M{
+			ActivatedKey:                true,
+			DeactivatedForDependencyKey: false,
+			ActivatedByKey:              caller,
+			ActivatedTimeKey:            time.Now(),
+		}},
+	)
+
+	for _, t := range tasksToActivate {
+		event.LogTaskActivated(t.Id, t.Execution, caller)
```
Comment (ablack12) [inline model/task/task.go:1005]: same question, can this be in loop at 992

File: model/task/task.go [hunk 8]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+	missingTaskMap := make(map[string]Task)
+	for _, t := range missingTasks {
+		missingTaskMap[t.Id] = t
+	}
+
+	// do a topological sort so we've adjudicated all a task's dependencies by the time
+	// we get up to it
+	sortedDependencies, err := topologicalSort(tasksDependingOnTheseTasks)
+	if err != nil {
+		return errors.Wrap(err, "can't do topological sort")
```
Comment (ablack12) [inline model/task/task.go:962]: this might not be worth wrapping, since the topologicalSort error is also just "problem with topological sort"

File: model/task/task.go [hunk 9]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
 				ScheduledTimeKey: utility.ZeroTime,
 			},
 		},
 	)
+	if err != nil {
+		return errors.Wrap(err, "problem deactivating tasks")
+	}
+
+	for _, t := range tasks {
+		event.LogTaskDeactivated(t.Id, t.Execution, caller)
```
Comment (ablack12) [inline model/task/task.go:1079]: In DeactivateDependencies you log regardless of error, we should be consistent about when we do this.

File: model/task/task.go [hunk 10]
```diff
@@ -1050,61 +1268,126 @@ func (t *Task) UpdateHeartbeat() error {
...
+			},
 		}},
-		bson.M{"$set": modifier},
+		bson.M{"$set": bson.M{PriorityKey: priority}},
 	)
+	if err != nil {
+		return errors.Wrap(err, "can't update priority")
+	}
 
 	event.LogTaskPriority(t.Id, t.Execution, user, priority)
```
Comment (ablack12) [inline model/task/task.go:1301]: Do we need to also log for the dependencies?

File: model/task_lifecycle.go [hunk 11]
```diff
@@ -41,12 +41,14 @@ func SetActiveState(taskId string, caller string, active bool) error {
...
 		// if the task is being activated and it doesn't override its dependencies
 		// activate the task's dependencies as well
+		tasksToActivate := []task.Task{}
 		if !t.OverrideDependencies {
-			for _, dep := range t.DependsOn {
-				if err = SetActiveState(dep.TaskId, caller, true); err != nil {
-					return errors.Wrapf(err, "error activating dependency for %v with id %v",
-						taskId, dep.TaskId)
-				}
+			deps, err := task.GetRecursiveDependenciesUp([]task.Task{*t}, nil)
```
Comment (ablack12) [inline model/task_lifecycle.go:46]: Will the result of GetRecursiveDependenciesUp include t? Or else should we add this at 44?

File: model/task_queue_service_dependency.go [hunk 12]
```diff
@@ -574,3 +550,27 @@ func (d *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
 
 	return nil
 }
+
+func checkUnmarkedBlockingTasks(t *task.Task, dependencyCaches map[string]task.Task) error {
+	catcher := grip.NewBasicCatcher()
+
+	blockingTasks, err := t.RefreshBlockedDependencies(dependencyCaches)
+	catcher.Add(errors.Wrap(err, "can't get blocking tasks"))
+	if err == nil && len(blockingTasks) > 0 {
```
Comment (ablack12) [inline model/task_queue_service_dependency.go:559]: You don't actually need to check len(blockingTasks) > 0, the for loop just won't do anything in that case

File: model/task_queue_service_dependency.go [hunk 13]
```diff
@@ -574,3 +550,27 @@ func (d *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
...
+		for _, tId := range blockingTasks {
+			tDb, err := task.FindOneId(tId)
+			catcher.Add(errors.Wrapf(err, "can't get task '%s'", tId))
+			if err == nil {
```
Comment (ablack12) [inline model/task_queue_service_dependency.go:563]: also check tDb != nil

File: model/task_queue_service_dependency.go [hunk 14]
```diff
@@ -574,3 +550,27 @@ func (d *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
...
+				catcher.Add(errors.Wrapf(UpdateBlockedDependencies(tDb), "can't update blocked dependencies for '%s'", tId))
+			}
+		}
+	}
+
+	blockingDeactivatedTasks, err := t.BlockedOnDeactivatedDependency(dependencyCaches)
+	catcher.Add(errors.Wrap(err, "can't get blocked status"))
+	if err == nil && len(blockingTasks) > 0 {
```
Comment (ablack12) [inline model/task_queue_service_dependency.go:571]: did you mean to check the length of blockingDeactivatedTasks?

File: model/task_queue_service_dependency.go [hunk 15]
```diff
@@ -574,3 +550,27 @@ func (d *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
 
 	return nil
 }
+
+func checkUnmarkedBlockingTasks(t *task.Task, dependencyCaches map[string]task.Task) error {
+	catcher := grip.NewBasicCatcher()
```
Comment (ablack12) [inline model/task_queue_service_dependency.go:555]: what's the reason for using the catcher here instead of just erroring at the point of error?

File: model/task/task.go [hunk 16]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+	for _, t := range tasks {
+		taskMap[t] = true
+	}
+
+	depTaskMap := make(map[string]bool)
+	tasksDependingOnTheseTasks, err := getRecursiveDependenciesDown(tasks, nil)
+	if err != nil {
+		return errors.Wrap(err, "can't get recursive dependencies down")
+	}
+	for _, t := range tasksDependingOnTheseTasks {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/task/task.go]: done

File: model/lifecycle.go [hunk 17]
```diff
@@ -82,80 +82,65 @@ func SetVersionActivation(versionId string, active bool, caller string) error {
...
 			// update all tasks if the caller is not evergreen.
-			_, err = task.UpdateAll(
-				bson.M{
-					task.BuildIdKey: bson.M{"$in": buildIds},
-					task.StatusKey:  evergreen.TaskUndispatched,
-				},
-				bson.M{"$set": bson.M{task.ActivatedKey: active, task.ActivatedByKey: caller}},
-			)
+			query = bson.M{
+				task.BuildIdKey: bson.M{"$in": buildIds},
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/lifecycle.go]: done

File: model/task/task.go [hunk 18]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+	return errors.Wrap(DeactivateDependencies(taskIDs, caller), "can't deactivate dependencies")
+}
+
+func DeactivateDependencies(tasks []string, caller string) error {
+	tasksDependingOnTheseTasks, err := getRecursiveDependenciesDown(tasks, nil)
+	if err != nil {
+		return errors.Wrap(err, "can't get recursive dependencies down")
+	}
+
+	tasksToUpdate := make([]Task, 0, len(tasksDependingOnTheseTasks))
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/task/task.go:1089]: `make([]Task, 0, len(tasksDependingOnTheseTasks))` produces a slice with a length of 0 and a capacity of len(tasksDependingOnTheseTasks) (https://tour.golang.org/moretypes/13). I think the idea is that if you know the length beforehand you can avoid reallocation of the underlying array as you fill it.

File: model/task/task.go [hunk 19]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+		}
+	}
+
+	if len(tasksToUpdate) == 0 {
+		return nil
+	}
+
+	taskIDsToUpdate := make([]string, 0, len(tasksToUpdate))
+	for _, t := range tasksToUpdate {
+		taskIDsToUpdate = append(taskIDsToUpdate, t.Id)
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/task/task.go:1102]: Is there a difference between doing two things in a loop or one in two?

[the remark this answers is not in the document]
Comment (ybrill) [reply model/task/task.go]: done

File: model/task/task.go [hunk 20]
```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
...
+		},
+		bson.M{"$set": bson.M{
+			ActivatedKey:                false,
+			DeactivatedForDependencyKey: true,
+			ScheduledTimeKey:            utility.ZeroTime,
+		}},
+	)
+
+	for _, t := range tasksToUpdate {
+		event.LogTaskDeactivated(t.Id, t.Execution, caller)
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/task/task.go:1114]: I think it would be confusing if we log the deactivation when it hasn't actually been deactivated

File: model/lifecycle.go (hunk 1 above)
Comment (ybrill) [reply model/lifecycle.go:110]: done

File: model/task/task.go (hunk 2 above)
Comment (ybrill) [reply model/task/task.go]: fixed

File: model/task/task.go (hunk 3 above)
Comment (ybrill) [reply model/task/task.go]: done

File: model/task/task.go (hunk 4 above)
Comment (ybrill) [reply model/task/task.go:908]: it looks to me like they both follow the same pattern: after the update was successful the event is logged.
Am I misunderstanding?

File: model/task/task.go (hunk 5 above)
Comment (ybrill) [reply model/task/task.go]: done

File: model/task/task.go (hunk 6 above)
Comment (ybrill) [reply model/task/task.go:974]: I thought the effect of having it would be the same as leaving it out (because `depTask.Activated` will be false when `!ok`)

File: model/task/task.go (hunk 8 above)
Comment (ybrill) [reply model/task/task.go]: changed to `WithStack`

File: model/task/task.go (hunk 7 above)
Comment (ybrill) [reply model/task/task.go:1004]: this is an extension of the previous disagreement about logging when the update wasn't successful.
FWIW, the logging we've been doing has always been after the [activation](https://github.com/evergreen-ci/evergreen/blob/59f50141ee548614ee33feabcdeefd5e9e79cd6d/model/task_lifecycle.go#L61)/[deactivation](https://github.com/evergreen-ci/evergreen/blob/59f50141ee548614ee33feabcdeefd5e9e79cd6d/model/task_lifecycle.go#L321) was successful

File: model/task/task.go (hunk 9 above)
Comment (ybrill) [reply model/task/task.go:1078]: I think this is the point I don't follow.
I thought DeactivateDependencies only logs when the update was successful.

File: model/task/task.go (hunk 10 above)
Comment (ybrill) [reply model/task/task.go]: good point!
Done

File: model/task_queue_service_dependency.go (hunk 12 above)
Comment (ybrill) [reply model/task_queue_service_dependency.go]: removed

File: model/task_queue_service_dependency.go (hunk 13 above)
Comment (ybrill) [reply model/task_queue_service_dependency.go]: code was removed

File: model/task_queue_service_dependency.go (hunk 14 above)
Comment (ybrill) [reply model/task_queue_service_dependency.go]: yes 😞

Comment (ybrill) [reply model/task_queue_service_dependency.go]: done

File: model/task_queue_service_dependency.go (hunk 15 above)
Comment (ybrill) [reply model/task_queue_service_dependency.go:555]: if the blocked check fails the activated check is still useful.

File: model/task_lifecycle.go (hunk 11 above)
Comment (ybrill) [reply model/task_lifecycle.go:39]: it won't include t, but t is included later if we aren't resetting the task.

File: model/task/task.go [hunk 21]
```diff
@@ -469,27 +472,19 @@ func (t *Task) RefreshBlockedDependencies(depCache map[string]Task) ([]string, e
...
 		if !ok {
-			grip.Error(message.Fields{
-				"task_id":     t.Id,
-				"dep_task_id": dep.TaskId,
-				"cache_size":  len(deps),
-				"operation":   "RefreshBlockedDependencies",
-				"message":     "cache failure",
-			})
-			continue
+			return nil, errors.Errorf("task '%s' is not in the cache", dep.TaskId)
```
Comment (ablack12) [inline model/task/task.go:484]: Are you confident this error isn't frequent?

File: model/task_lifecycle.go [hunk 22]
```diff
@@ -30,35 +30,29 @@ type StatusChanges struct {
...
 		if t.DispatchTime != utility.ZeroTime && t.Status == evergreen.TaskUndispatched {
-			if err = resetTask(t.Id, caller); err != nil {
+			if err := resetTask(t.Id, caller); err != nil {
 				return errors.Wrap(err, "error resetting task")
 			}
 		} else {
-			if err = t.ActivateTask(caller); err != nil {
-				return errors.Wrap(err, "error while activating task")
-			}
-			event.LogTaskActivated(taskId, t.Execution, caller)
```
Comment (ablack12) [inline model/task_lifecycle.go:61]: Why do we remove these logs?

File: model/task/task.go (hunk 9 above)
Comment (ablack12) [reply model/task/task.go:1078]: Ah in DeactivateDependencies the error isn't checked until after the loop that logs tasks being deactivated.

File: model/task/task.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (ablack12) [reply model/task/task.go:1114]: to be clear, we should check that the update didn't error at 1112 instead of 1117

File: model/task_lifecycle.go (hunk 22 above)
Comment (ybrill) [reply model/task_lifecycle.go:61]: The event is logged at the end of `task.ActivateTasks`. 
Consolidating logging the event into the function that makes the event makes sure developers don't forget to log the event.

File: model/task/task.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (ybrill) [reply model/task/task.go:1117]: yes. I get it now.
done

**Comment by ybrill** [comment]
I tested out these changes in staging and noticed the build cache wasn't getting updated for dependencies activated/deactivated.
I made changes to update the build cache. @ablack12 would you mind looking over the latest changes when you have a chance?

File: model/lifecycle.go [hunk 23]
```diff
@@ -238,6 +238,61 @@ func MarkVersionCompleted(versionId string, finishTime time.Time, updates *Statu
...
 }
 
+func SetTaskPriority(t task.Task, priority int64, caller string) error {
+	depTasks, err := task.GetRecursiveDependenciesUp([]task.Task{t}, nil)
+	if err != nil {
+		return errors.Wrap(err, "error getting task dependencies")
+	}
+
+	ids := []string{t.Id}
+	ids = append(ids, t.ExecutionTasks...)
```
Comment (ablack12) [inline model/lifecycle.go:248]: nit: elsewhere you did `ids := append([]string{t.Id}, t.ExecutionTasks...)`

File: model/task/task.go [hunk 24]
```diff
@@ -1268,56 +1281,38 @@ func (t *Task) UpdateHeartbeat() error {
...
-func (t *Task) SetPriority(priority int64, user string) error {
-	t.Priority = priority
-	depTasks, err := GetRecursiveDependenciesUp([]Task{*t}, nil)
-	if err != nil {
-		return errors.Wrap(err, "error getting task dependencies")
-	}
+// Blacklist sets the priority of a task so it will never run.
+// It also deactivates the task and any tasks that depend on it.
+// The build cache is not updated
+func (t *Task) Blacklist(user string) ([]Task, error) {
```
Comment (ablack12) [inline model/task/task.go:1287]: How come this function can't just call SetTaskPriority? If SetTaskPriority won't properly handle the blacklist case, then it should probably just be modified to handle it, or shouldn't accept it.

File: model/task_lifecycle.go [hunk 25]
```diff
@@ -306,16 +308,9 @@ func DeactivatePreviousTasks(t *task.Task, caller string) error {
 			continue
 		}
 
-		err = SetActiveState(&t, caller, false)
-		if err != nil {
+		if err = SetActiveState(&t, caller, false); err != nil {
 			return err
 		}
-		// update the cached version of the task, in its build document to be deactivated
-		if !t.IsPartOfDisplay() {
```
Comment (ablack12) [inline model/task_lifecycle.go:323]: Why did you remove this?

File: rest/data/task.go [hunk 26]
```diff
@@ -146,7 +147,10 @@ func (tc *DBTaskConnector) FindTasksByProjectAndCommit(projectId, commitHash, ta
 // SetTaskPriority changes the priority value of a task using a call to the
 // service layer function.
 func (tc *DBTaskConnector) SetTaskPriority(t *task.Task, user string, priority int64) error {
-	err := t.SetPriority(priority, user)
+	if t == nil {
+		return errors.New("task cannot be nil")
+	}
+	err := model.SetTaskPriority(*t, priority, user)
```
Comment (ablack12) [inline rest/data/task.go:153]: Nit: just return here

File: model/task_queue_service_dependency.go [hunk 27]
```diff
@@ -565,7 +566,12 @@ func checkUnmarkedBlockingTasks(t *task.Task, dependencyCaches map[string]task.T
 	blockingDeactivatedTasks, err := t.BlockedOnDeactivatedDependency(dependencyCaches)
 	catcher.Add(errors.Wrap(err, "can't get blocked status"))
 	if err == nil && len(blockingDeactivatedTasks) > 0 {
-		catcher.Add(errors.Wrapf(task.DeactivateDependencies(blockingDeactivatedTasks, evergreen.DefaultTaskActivator+".dispatcher"), "can't deactivate task '%s'", t.Id))
+		var deactivatedDependencies []task.Task
+		deactivatedDependencies, err = task.DeactivateDependencies(blockingDeactivatedTasks, evergreen.DefaultTaskActivator+".dispatcher")
+		catcher.Add(err)
+		if err != nil {
```
Comment (ablack12) [inline model/task_queue_service_dependency.go:572]: Should this be err == nil, as in there was no problem deactivating the dependencies?

File: model/build/task_cache.go [hunk 28]
```diff
@@ -99,6 +101,51 @@ func SetCachedTaskActivated(buildId, taskId string, active bool) error {
...
+	for _, t := range tasks {
+		taskID := t.Id
+		if t.IsPartOfDisplay() {
+			taskID = t.DisplayTask.Id
+		}
+
+		if _, ok := buildMap[t.BuildId]; !ok {
+			buildMap[t.BuildId] = make(map[string]bool)
+		}
+		buildMap[t.BuildId][taskID] = true
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/build/task_cache.go:116]: That was how I had it originally, but I realized for display tasks the same display task will come up multiple times if more than one of its execution tasks is (de)activated.

File: model/lifecycle.go [hunk 29]
```diff
@@ -238,6 +238,61 @@ func MarkVersionCompleted(versionId string, finishTime time.Time, updates *Statu
...
+	depIDs := make([]string, 0, len(depTasks))
+	for _, task := range depTasks {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/lifecycle.go]: done

File: model/lifecycle.go (hunk 23 above)
Comment (ybrill) [reply model/lifecycle.go]: done

File: rest/data/task.go (hunk 26 above)
Comment (ybrill) [reply rest/data/task.go]: done

File: model/task_queue_service_dependency.go (hunk 27 above)
Comment (ybrill) [reply model/task_queue_service_dependency.go]: yes.
Fixed

File: model/task_lifecycle.go (hunk 25 above)
Comment (ybrill) [reply model/task_lifecycle.go:323]: I was trying to consolidate the places where we update the build cache tasks' activation to the places where we modify the actual tasks' activations. 
Now SetActiveState handles setting the build cache itself. 

I really would have wanted to do it inside the task package (in the functions that (de)activate tasks) but it made an import cycle (because the build package imports the task package). Instead, I settled for updating both the task and build cache from the model package (task_lifecycle and lifecycle).

File: model/task/task.go (hunk 24 above)
Comment (ybrill) [reply model/task/task.go:1287]: 😞 This is a good question.
I wanted to move SetTaskPriority into the model package so it could update the build cache (to [consolidate](https://github.com/evergreen-ci/evergreen/pull/3689#discussion_r442276859) the places where we update the build cache) but the commitqueue package couldn't import the model package. The commit queue can import the task and build packages, though, so I made this function to do just what the commit queue needed for the task (even though it's a subset of what  SetTaskPriority already does and is a violation of DRY). The commit queue calls another function in the build package to update the build cache.
